cocos2d/core/renderer/RendererWebGL.js

```diff
@@ -23,8 +22,104 @@
  THE SOFTWARE.
  ****************************************************************************/
 
+// Internal variables
+    // Batching general informations
+var _batchedInfo = {
+        // The batched texture, all batching element should have the same texture
+        texture: null,
+        // The batched blend source, all batching element should have the same blend source
+        blendSrc: null,
+        // The batched blend destination, all batching element should have the same blend destination
+        blendDst: null,
+        // The batched shader, all batching element should have the same shader
+        shader: null
+    },
+
+    _indexBuffer = null,
+    _vertexBuffer = null,
+    // Total vertex size
+    _maxVertexSize = 0,
+    // Current batching vertex size
+    _batchingSize = 0,
+    // Current batching index size
+    _indexSize = 0,
+    // Float size per vertex
+    _sizePerVertex = 6,
+    // buffer data and views
+    _vertexData = null,
+    _vertexDataSize = 0,
+    _vertexDataF32 = null,
+    _vertexDataUI32 = null,
+    _indexData = null,
+    _prevIndexSize = false,
+    _pureQuad = true,
+    _IS_IOS = false;
+
+
+// Inspired from @Heishe's gotta-batch-them-all branch
+// https://github.com/Talisca/cocos2d-html5/commit/de731f16414eb9bcaa20480006897ca6576d362c
+function updateBuffer (numQuads) {
+    var gl = cc._renderContext;
+    // Update index buffer size
+    if (_indexBuffer) {
+        gl.bindBuffer(gl.ELEMENT_ARRAY_BUFFER, _indexBuffer);
+        _indexData = new Uint16Array(numQuads * 6);
+        var currentQuad = 0;
+        for (var i = 0, len = numQuads * 6; i < len; i += 6) {
+            _indexData[i] = currentQuad + 0;
+            _indexData[i + 1] = currentQuad + 1;
+            _indexData[i + 2] = currentQuad + 2;
+            _indexData[i + 3] = currentQuad + 1;
+            _indexData[i + 4] = currentQuad + 2;
+            _indexData[i + 5] = currentQuad + 3;
+            currentQuad += 4;
+        }
```

Comment:
Initialize with quad data because most case there won't be triangle data be batched, but when it does, the index buffer will get updated correctly. Otherwise, we don't need to update or upload index buffer.